skyportal/handlers/api/shift.py

```diff
@@ -0,0 +1,146 @@
+from ..base import BaseHandler
+from ...models import (
+    DBSession,
+    Group,
+    Shift,
+)
+
+
+class ShiftHandler(BaseHandler):
+    @permissions(["Upload data"])
+    def post(self):
+        """
+        ---
+        description: Add a new shift
+        tags:
+          - shift
+        requestBody:
+          content:
+            application/json:
+              schema: ShiftNoID
+        responses:
+          200:
+            content:
+              application/json:
+                schema:
+                  allOf:
+                    - $ref: '#/components/schemas/Success'
+                    - type: object
+                      properties:
+                        data:
+                          type: object
+                          properties:
+                            id:
+                              type: integer
+                              description: New Shift
+          400:
+            content:
+              application/json:
+                schema: Error
+        """
+        data = self.get_json()
+
+        schema = Shift.__schema__()
+
+        try:
+            shift = schema.load(data)
+        except ValidationError as exc:
+            return self.error(
+                f"Invalid/missing parameters: {exc.normalized_messages()}"
+            )
+
+        DBSession().add(shift)
+        self.verify_and_commit()
+
+        self.push_all(action="skyportal/REFRESH_SHIFTS")
+        return self.success(data={"id": shift.id})
+
+    @auth_or_token
+    def get(self, group_id=None):
+        """
+        ---
+        description: Retrieve shifts
+        tags:
+          - shifts
+        parameters:
+          - in: path
+            name: group_id
+            required: false
+            schema:
+              type: integer
+        responses:
+          200:
+            content:
+              application/json:
+                schema: ArrayOfShifts
+          400:
+            content:
+              application/json:
+                schema: Error
+        """
+        if group_id is not None:
+            shifts = (
+                Shift.query_records_accessible_by(
+                    self.current_user,
+                    mode="read",
+                    options=[joinedload(Shift.group).joinedload(Group.users)],
+                )
+                .filter(Shift.group_id == group_id)
+                .order_by(Shift.start_date.asc())
+                .all()
+            )
+        else:
+            shifts = (
+                Shift.query_records_accessible_by(
+                    self.current_user,
+                    mode="read",
+                    options=[joinedload(Shift.group).joinedload(Group.users)],
+                )
+                .order_by(Shift.start_date.asc())
+                .all()
+            )
+        self.verify_and_commit()
+        return self.success(data=shifts)
+
+    @permissions(["Upload data"])
```

Comment:
should be a permissions to "Manage shifts"

Reply:
added